soh/soh/Enhancements/randomizer/3drando/item_pool.cpp

```diff
@@ -1346,31 +1336,30 @@ void GenerateItemPool() {
     ReplaceMaxItem(RG_DOUBLE_DEFENSE, 0);
   }
 
-  //this feels ugly and there's probably a better way, but
-  //it replaces random junk with pending junk.
-  bool junkSet;
-  for (RandomizerGet pendingJunk : PendingJunkPool) {
-    junkSet = false;
-    for (RandomizerGet& item : ItemPool) {
-      for (RandomizerGet junk : JunkPoolItems) {
-        if (item == junk && item != RG_HUGE_RUPEE && item != RG_DEKU_NUTS_10) {
-          item = pendingJunk;
-          junkSet = true;
+  std::erase(ItemPool, RG_NONE);
+
+  if (ItemPool.size() < ctx->allLocations.size()) {
+    Shuffle(PendingJunkPool);
+    size_t junkNeeded = std::min(PendingJunkPool.size(), ctx->allLocations.size() - ItemPool.size());
+    ItemPool.insert(ItemPool.end(), PendingJunkPool.begin(), PendingJunkPool.begin() + junkNeeded);
+    PendingJunkPool.erase(PendingJunkPool.begin(), PendingJunkPool.begin() + junkNeeded);
+  } else if (ItemPool.size() > ctx->allLocations.size()) {
+    // RANDOTODO: all junk should be put in PendingJunkPool so this is never needed
+    size_t remove = ItemPool.size() - ctx->allLocations.size();
+    for (size_t i = 0; remove > 0 && i < ItemPool.size(); i++) {
+      for (size_t j = 0; j < JunkPoolItems.size(); j++) {
+        if (ItemPool[i] == JunkPoolItems[j]) {
+          ItemPool[i] = RG_NONE;
+          remove--;
           break;
         }
       }
-      if (junkSet) break;
     }
+    std::erase(ItemPool, RG_NONE);
   }
-  PendingJunkPool.clear();
 
   // RANDOTODO: Ideally this should be checking for equality, but that is not currently the case and has never been
   // the case, and isn't even currently the case in the 3drando repo we inherited this from years ago, so it may
   // be a large undertaking to fix.
   assert(ItemPool.size() <= ctx->allLocations.size() || !"Item Pool larger than Location Pool");
```

Comment:
with this PR it might be suitable to return a status code so that generation can fail. This would be a situation when trying 100 triforce pieces & overworld keys & boss souls & npc souls in future so that item pool size exceeds available locations. Guess it's less of an issue for beatable only